std/string.d

```diff
@@ -2962,39 +2962,188 @@ S leftJustify(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
     Right justify $(D s) in a field $(D width) characters wide. $(D fillChar)
     is the character that will be used to fill up the space in the field that
     $(D s) doesn't fill.
+
+    Params:
+        s = string
+        width = minimum field width
+        fillChar = used to pad end up to $(D width) characters
+
+    Returns:
+        GC allocated string
+
+    See_Also:
+        $(LREF rightJustifier), which does not allocate
   +/
-S rightJustify(S)(S s, size_t width, dchar fillChar = ' ') @trusted pure
+S rightJustify(S)(S s, size_t width, dchar fillChar = ' ')
     if (isSomeString!S)
 {
-    import std.utf : canSearchInCodeUnits;
-    import std.conv : to;
+    import std.array;
+    return rightJustifier(s, width, fillChar).array;
+}
 
-    alias C = ElementEncodingType!S;
+/++
+    Right justify $(D s) in a field $(D width) characters wide. $(D fillChar)
+    is the character that will be used to fill up the space in the field that
+    $(D s) doesn't fill.
 
-    if (canSearchInCodeUnits!C(fillChar))
-    {
-        immutable len = s.walkLength();
-        if (len >= width)
-            return s;
+    Params:
+        s = string or forward range of characters
+        width = minimum field width
+        fillChar = used to pad end up to $(D width) characters
 
-        auto retval = new Unqual!C[width - len + s.length];
-        retval[0 .. $ - s.length] = cast(C)fillChar;
-        retval[$ - s.length .. $] = s[];
-        return cast(S)retval;
+    Returns:
+        a lazy range of the right justified result
+
+    See_Also:
+        $(LREF leftJustifier)
+  +/
+
+auto rightJustifier(Range)(Range r, size_t width, dchar fillChar = ' ')
+    if (isForwardRange!Range && isSomeChar!(ElementEncodingType!Range))
+{
+    alias C = Unqual!(ElementEncodingType!Range);
+
+    static if (C.sizeof == 1)
+    {
+        import std.utf : byDchar, byChar;
+        return rightJustifier(r.byDchar, width, fillChar).byChar;
     }
-    else
+    else static if (C.sizeof == 2)
     {
-        auto dstr = to!dstring(s);
-        if (dstr.length >= width)
-            return s;
+        import std.utf : byDchar, byWchar;
+        return rightJustifier(r.byDchar, width, fillChar).byWchar;
+    }
+    else static if (C.sizeof == 4)
+    {
+        static struct Result
+        {
+          private:
+            Range _input;
+            size_t _width;
+            alias _width nfill;       // number of fill characters to prepend
+            dchar _fillChar;
+            bool inited;
+
+            // Lazy initialization so constructor is trivial and cannot fail
+            void initialize()
+            {
+                // Replace _width with nfill
+                // (use alias instead of union because CTFE cannot deal with unions)
+                assert(_width);
+                static if (hasLength!Range)
+                {
+                    auto len = _input.length;
+                    nfill = (_width > len) ? _width - len : 0;
```

Comment:
Eh, I'd have to add another import statement, winding up with the same number of lines. Doesn't seem worth it.